napari/_vispy/layers/labels.py

```diff
@@ -164,11 +173,46 @@ def hash2d_set(key, value, keys, values, empty_val=0):
             raise OverflowError('too many labels')
         key += 1
         pos = idx_to_2D(key, keys.shape)
-    keys[pos] = key
+    keys[pos] = initial_key
     values[pos] = value
 
 
-def build_textures_from_dict(color_dict, empty_val=0, shape=(1000, 1000)):
+def _get_shape_from_dict(color_dict):
+    size = len(color_dict) * 4
+    # I think that hash table size should be at least 4 times
+    # bigger than the number of labels to avoid collisions
+    for i, prime in enumerate(PRIME_NUM_TABLE[:-1]):
+        if prime * prime > size:
+            return prime, prime
+        if prime * PRIME_NUM_TABLE[i + 1] > size:
+            return PRIME_NUM_TABLE[i + 1], prime
+
+    if size > PRIME_NUM_TABLE[-1] * PRIME_NUM_TABLE[-1]:
+        raise OverflowError('too many labels')
+    return PRIME_NUM_TABLE[-1], PRIME_NUM_TABLE[-1]
+
+
+def get_shape_from_dict(color_dict):
+    global MAX_TEXTURE_SIZE
+    if MAX_TEXTURE_SIZE is None:
+        MAX_TEXTURE_SIZE = get_max_texture_sizes()
+
+    shape = _get_shape_from_dict(color_dict)
+
+    if MAX_TEXTURE_SIZE is not None and (
+        shape[0] > MAX_TEXTURE_SIZE[0] or shape[1] > MAX_TEXTURE_SIZE[1]
+    ):
+        raise OverflowError(
+            f'Too many labels. GPU does not support textures of this size.'
+            f' Requested size is {shape[0]}x{shape[1]}, but maximum supported'
+            f' size is {MAX_TEXTURE_SIZE[0]}x{MAX_TEXTURE_SIZE[1]}'
+        )
+    return shape
```

Comment:
@brisvag what did you think about this implementation 
(based on https://github.com/napari/napari/pull/6179#issuecomment-1688051007)

Reply:
I think this makes sense. Why exactlu are you using a global variable here?

Reply:
Ok. I have changed to limit texture size to use the size that will work in both 2d and 3d mode.

Reply:
> Performance. But now I check the code, and I need to fix this, as it is the max size for 2d and 3d not the max size of the texture dimension.

Ah true!